blockmanager_test.go

```diff
@@ -0,0 +1,230 @@
+	}
+
+	// Keep track of the filter headers and block headers. Since the
+	// genesis headers are written automatically when the store is created,
+	// we query it to add to the slices.
+	genesisBlockHeader, _, err := hdrStore.ChainTip()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var blockHeaders []headerfs.BlockHeader
+	blockHeaders = append(blockHeaders, headerfs.BlockHeader{
+		BlockHeader: genesisBlockHeader,
+		Height:      0,
+	})
+
+	genesisFilterHeader, _, err := cfStore.ChainTip()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var cfHeaders []chainhash.Hash
+	cfHeaders = append(cfHeaders, *genesisFilterHeader)
+
+	// Also keep track of the current filter header. We use this to
+	// calculate the next filter header, as it commits to the previous.
+	currentCFHeader := *genesisFilterHeader
+
+	// checkpoints will be the checkpoints passed to
+	// getCheckpointedCFHeaders.
+	var checkpoints []*chainhash.Hash
+
+	maxHeight := 50 * uint32(wire.CFCheckptInterval)
+	for height := uint32(1); height <= maxHeight; height++ {
+		header := heightToHeader(height)
+		blockHeader := headerfs.BlockHeader{
+			BlockHeader: header,
+			Height:      height,
+		}
+
+		blockHeaders = append(blockHeaders, blockHeader)
+
+		// It doesn't really matter what filter the filter header
+		// commit to, so just use the height as a nonce.
+		filterHash := chainhash.Hash{}
+		binary.BigEndian.PutUint32(filterHash[:], height)
+		currentCFHeader = chainhash.DoubleHashH(
+			append(filterHash[:], currentCFHeader[:]...),
+		)
+		cfHeaders = append(cfHeaders, filterHash)
+
+		// Each interval we must record a checkpoint.
+		if height > 0 && height%wire.CFCheckptInterval == 0 {
+			// We must make a copy of the current header.
+			cfh := currentCFHeader
+			checkpoints = append(checkpoints, &cfh)
+		}
+
+	}
+
+	// Write all block headers but the genesis, since it is already in the
+	// store.
+	if err = hdrStore.WriteHeaders(blockHeaders[1:]...); err != nil {
+		t.Fatalf("Error writing batch of headers: %s", err)
+	}
+
+	var wg sync.WaitGroup
+
+	// Set up a chain service with a custom query batch method.
+	cs := &ChainService{
+		BlockHeaders: hdrStore,
+	}
+	cs.queryBatch = func(msgs []wire.Message, f func(*ServerPeer,
+		wire.Message, wire.Message) bool, q <-chan struct{},
+		qo ...QueryOption) {
+
+		// Craft response for each message.
+		for _, msg := range msgs {
+			// Only GetCFHeaders expected.
+			q, ok := msg.(*wire.MsgGetCFHeaders)
+			if !ok {
+				t.Fatalf("got unexpected message %T", msg)
+			}
+
+			// The start height must be set to a checkpoint
+			// height+1.
+			if q.StartHeight%wire.CFCheckptInterval != 1 {
+				t.Fatalf("unexpexted start height %v",
+					q.StartHeight)
+			}
+
+			var prevFilterHeader chainhash.Hash
+			switch q.StartHeight {
+
+			// If the start height is 1 the prevFilterHeader is set
+			// to the genesis header.
+			case 1:
+				prevFilterHeader = *genesisFilterHeader
+
+			// Otherwise we use one of the created checkpoints.
```

Comment:
FWIW, this only works as is, since we only request 1k headers at a time. If we started to request 2k headers at a time, then this would break. Even then, it assumes that it will only ever request headers at 1k boundary. 


IMO, this is fine for this test and particular case, but when we do the full on p2p testing harness, we'll want to make sure it can handle any sort of request (we'd basically just proxy a set of pseudo-nodes using btcd's RPC server).